tools: don't use global Buffer and add linter rule for it

silverwind opened the pull request:
This does a few things to the linter configuration:

- instead of relying on the predefined `node` environment, specify all our globals manually and as read-only.
- enable [`no-undef`](http://eslint.org/docs/rules/no-undef) to error on undefined variable access.
- add `globalReturn` which allows to return from the top scope. This was implied by the `node` option before.

A number of tests intentionally use undefined globals to trigger an error or set globals (in the `vm` tests). I had to exclude those tests with an `eslint-disable` comment. If these comments are too noisy, we could alternatively disable that rule for all tests.

I might have missed a few globals from the v8 environment. `Object.keys(global)` doesn't list many of them, presumably because they are unenumerable. Any better ideas on how to discover globals?

cc: @Fishrock123 @domenic

silverwind commented:
@Fishrock123 what's the plan exactly with Buffer usage mentioned in https://github.com/nodejs/io.js/pull/1770? Should it error as soon as `Buffer` is used inside `lib`?

silverwind commented on test/parallel/test-cluster-worker-kill.js:
```diff
@@ -106,16 +106,12 @@ function checkResults(expected_results, results) {
         expected = expected_results[k];
 
-    if (typeof expected === 'function') {
-      expected(r[k]);
```
This is an actual error with the undefined `r` variable here. Because the condition was never met, I removed the whole `if` block.

silverwind commented on test/parallel/test-cluster-worker-exit.js:
```diff
@@ -106,17 +106,13 @@ function checkResults(expected_results, results) {
         expected = expected_results[k];
 
-    if (typeof expected === 'function') {
-      expected(r[k]);
```
This is an actual error with the undefined `r` variable here. Because the condition was never met, I removed the whole `if` block.

yosuke-furukawa commented on .eslintrc:
```diff
@@ -9,6 +6,7 @@ ecmaFeatures:
   binaryLiterals: true
   generators: true
   forOf: true
+  globalReturn: true
```
why is `globalReturn` needed?

yosuke-furukawa replied:
Oh, I didn't see your comment. I understand that.

yosuke-furukawa commented on .eslintrc:
```diff
@@ -29,6 +27,8 @@ rules:
   ## require valid typeof compared string like typeof foo === 'strnig'
   valid-typeof: 2
+  ## disallow use of undefined variables (globals)
+  no-undef: 2
```
no-undef is not `Possible Errors` category,
`Variables` category, https://github.com/eslint/eslint/tree/master/docs/rules#variables

silverwind replied:
Yeah, sorry about that, will fix.

Fishrock123 commented:
> @fishrock123 what's the plan exactly with Buffer usage mentioned in #1770? Should it error as soon as Buffer is used inside lib?

Probably? How else would it work?

Fishrock123 commented on test/message/nexttick_throw.js:
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable no-undef */
```
Perhaps disable for all tests?

silverwind replied:
Yeah, I think it's better that way, less noisy comments all over the tests, will work something out.

rlidwka replied:
`no-undef` is one of the few rules that do something good (as opposed to the most eslint rules that do nothing except for enforcing a ~~religion~~ code style). I don't think it should be disabled for all tests, because it could catch real errors.

What about explicitly declaring globals in every file?

```js
/* globals: undefined_reference_error_maker */
```

I don't remember the exact syntax, but I think you can add globals on per-file basis.

silverwind replied:
I agree generally to using `no-undef`, but it is pretty common to use undefined globals in tests to trigger an error. With the rule enabled, one would have to research how to configure the linter when writing a test, which I'd like to avoid.

Using a `global` comment is pretty equivalent to disabling the rule, both require additional ugly comments. I think we're fine with just disabling `no-undef` in the test directory.

rlidwka replied:
maybe ask people to trigger an error differently?

my favorite way is `[]()`

silverwind commented:
@Fishrock123 I'll just comment out the Buffer global in `.eslintrc` to indicate it's there but not allowed to be used directly. We want to use `require('buffer').Buffer` in place of it, right?

Fishrock123 commented:
@silverwind correct.

silverwind commented:
Rebased and disabled `no-undef` for tests, also moved the rule to the correct category in .eslintrc.

@Fishrock123 There are 83 occasions where `Buffer` is used directly in `lib`. Should I go for it and replace them with a non-global?

silverwind commented:
@Fishrock123 I'll follow up with a port of https://github.com/joyent/node/commit/523929c9272a53c9429616564a45f2af59670e47 in another PR. Does this one LGTY?

Fishrock123 commented:
@silverwind I'd prefer if you did the commits in this order: fix tests, apply full remove-global-buffer patch, add linting.

That being said, this still doesn't seem to catch global Buffer use with `./iojs tools/eslint/bin/eslint.js src lib test --reset`?

silverwind commented:
Well I suppose I can split these up into multiple commits and add the Buffer thing too in its own commit.

silverwind commented:
Alright, split up in 3 commits now. The circular depedency described in https://github.com/joyent/node/pull/8603#issuecomment-60297429 didn't happen, so no workaround was needed.

silverwind commented:
CI: https://jenkins-iojs.nodesource.com/view/iojs/job/iojs+any-pr+multi/728/

silverwind commented:
Updated and cleaned up the list of globals, which is now sorted in categories. @Fishrock123 PTAL

targos commented on .eslintrc:
```diff
@@ -63,35 +56,75 @@ rules:
+  __dirname         : false
+  global            : false
+  GLOBAL            : false
+  root              : false
```
`GLOBAL` and `root` are going to be deprecated (#1838). Are they still being used in the code base ? If not they do not need to be here.

thefourtheye replied:
But they ll still be there till removal, right?

silverwind replied:
They're not used and no one sane would use them, so good call on removing them.

silverwind commented:
Filed https://github.com/sindresorhus/globals/pull/33 for the missing `DataView` builtin in eslint, until then we probably have to keep it listed for correctness (it's only used in one test).

silverwind commented:
Also filed https://github.com/eslint/eslint/issues/2657 for a simpler way to undefine globals.

silverwind commented:
Don't review yet, I'll try creating a custom rule for the `Buffer` case, which would drastically simplify the current changes to globals.

silverwind commented:
Here's my current custom rule, which @xjamundx graciously provided:

````js
"use strict";

module.exports = function(context) {
  var required = false;
  return {
    CallExpression: function(node) {
      if (node.callee.type === 'Identifier' &&
          node.callee.name === 'require') {
        if (node.arguments.length && node.arguments[0].value === 'buffer') {
          required = true;
        }
      }
    },
    Identifier: function(node) {
      if (!required && node.name === 'Buffer') {
        context.report(node, "Make sure to require('buffer').Buffer before using Buffer");
      }
    }
  }
}
````

It's *almost* working, except that it does detect an error on the require line itself:

````js
const Buffer = require('buffer').Buffer;
//    ^--- errors here
````

Any suggestions?

rlidwka commented:
I think your suggestion in eslint/eslint#2657 is much better than this headache with custom rules. Maybe re-open it?

gcochard commented:
@silverwind, I would be happy to help out with debugging the custom rule if you'd like. 

I think right now the `Identifier` event is being emitted before the `CallExpression`. A *very* hack way to work through this is to push all `Identifier` nodes to an array and then add a hook to `Program:exit` and report all of them if you don't detect the require.

silverwind commented:
Yeah, I think that `Program:Exit` function is on track (as used in [`no-undef`](https://github.com/eslint/eslint/blob/c83d1e3e34cf516cf5a000d6dc94f860c3e5c7a9/lib/rules/no-undef.js#L67)).

I think a clean way to solve this could be to iterate `globalScope.through` and `scope.variables` and error if we encounter `Buffer`, much like `no-undef` already does.

silverwind commented:
I think I got it working, how does this one look to you @gcochard 

```js
'use strict';

const message = "Make sure to require('buffer').Buffer before using Buffer";

module.exports = function(context) {
  return {
    "Program:exit": function() {
      context.getScope().through.some(function(ref) {
        if (ref.identifier.name === 'Buffer')
            context.report(ref.identifier, message);
      });
    }
  }
}
```

gcochard commented:
Code looks good, I would make the error message more explicit that you need to initialize the `Buffer` var/const with the results of the `require` call.

gcochard commented:
I would also check that it does report an error when you use the global Buffer before requiring and defining your local const, given that const is not hoisted.

silverwind commented:
> use the global Buffer before requiring

Good point. But that seems like a candidate for a builtin eslint rule. I found `http://eslint.org/docs/rules/no-use-before-define` but it's not tailored to `let`and `const` it seems.

gcochard commented:
It doesn't need to be tailored for `let` or `const`, since using them before definition would be reported as a syntax error by the parser. It's a hard use-case because it's a global but you want to discourage use of it.

silverwind commented:
@gcochard Indeed the parser will give a `ReferenceError` before use. We run `make test` before each push, so I don't think the rule needs any more checks (I wouldn't even know how to do them). Thanks again!

gcochard commented:
Np, happy to help :+1:.

silverwind commented:
Up for review again!

trevnorris commented:
I'm realizing right now that unless we make the exports Buffer property read only it can still be overridden just as easily by users. So what's the point of forcing it to be required in all files?

silverwind commented:
I don't see writable Buffer as an problem really. It could allow for a custom Buffer implementation and a lot of things in node can be overridden right now anyways.

As for the reason of this change, there's a series of issues/prs that lead to it:

https://github.com/joyent/node/issues/8588
https://github.com/joyent/node/pull/8603
https://github.com/nodejs/io.js/pull/1770

trevnorris commented:
Ah yes. I forgot about the REPL issue. Thanks for the reminder.

I don't see the same change here for `assert.js` as in the final commit in https://github.com/joyent/node/commit/523929c9272a53c9429616564a45f2af59670e47 (using `b.Buffer` instead). Has that issue been fixed?

silverwind commented:
Yes, I think so. Tests were passing so I assumed it was fixed.

Fishrock123 commented:
@trevnorris I looked at that while attempting to port the original patch... I'm not sure why that was done in the first place?

Oh, found it here: https://github.com/joyent/node/pull/8603#issuecomment-60297429 Looks like there was a circular dependancy issue?

silverwind commented:
@Fishrock123 @trevnorris @vkurchatkin I don't understand why this race condition happened last time, but it seems to be gone now, all tests are passing so I assume it must've been fixed in the meanwhile.

I rebased the PR and did a minor correction to the `require-buffer` rule (switched useless `some` for `forEach`).

silverwind commented:
Did another small correction to the printed message. Please review.

trevnorris commented on lib/buffer.js:
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable require-buffer */
 'use strict';
```
I thought `'use strict'` had to go on the first line of either the file or function. Does that exclude comments?

/cc @domenic

trevnorris replied:
nm. see it doesn't matter.

rlidwka replied:
I'm pretty sure it does exclude comments, considering the amount of files with copyright hats at the very top.

trevnorris commented on test/pummel/test-net-timeout.js:
```diff
@@ -79,7 +79,7 @@ process.on('exit', function() {
   assert.ok(timeouttime != null);
 
-  diff = timeouttime - starttime;
+  var diff = timeouttime - starttime;
```
Strange that both common and `'use strict'` didn't throw an error for this variable that wasn't previously declared. Glad it's fixed, but bothers me that wasn't caught before. Any thoughts as to why?

silverwind replied:
Strange indeed. If I run this test manually on master without the framework, I see a ReferenceError. Not too familar with pummel tests, how are they executed?

trevnorris replied:
Same way all the other tests are run. `test.py pummel`

silverwind replied:
`make test-pummel` shows this error (along with some others). It seems these tests aren't ran as part of `make test` at all.

trevnorris replied:
ah yeah. forgot about that slight nuance. :P 

Reason they're not run with all the others is because they have the potential to run a long time, or overload the systems resources.

trevnorris commented on tools/eslint-rules/require-buffer.js:
```diff
@@ -0,0 +1,15 @@
+'use strict';
+
+const msg = "Use const Buffer = require('buffer').Buffer; on top of this file";
```
nit: s/on top of/at the beginning of/

thefourtheye replied:
Also, we use `'` for strings, right? Is it okay here?

silverwind replied:
To avoid escaping it's okay (There's a linter rule for it).

trevnorris commented:
Have one nit, but LGTM.

thefourtheye commented on tools/eslint-rules/require-buffer.js:
```diff
@@ -0,0 +1,15 @@
+
+module.exports = function(context) {
+  return {
+    "Program:exit": function() {
```
Same as above ^^

silverwind replied:
I'll fix that.

trevnorris commented:
Great thanks. I'm cool with this change.

silverwind commented:
Thanks! Landed in https://github.com/nodejs/io.js/commit/ff8202c6f475973c8674e61c794b77e1b072144b .. https://github.com/nodejs/io.js/commit/6e4d30286d7305335b14762271a25f5dbc5c63c0.
